### PR TITLE
[Bugfix] Fix FlashInfer sparse mask on small sequences

### PR DESCRIPTION
**Pull request by rick-heig**
Sequences smaller than sparse block block size would result in block_row_sizes/block_col_sizes being empty and accessing them with [-1] to append the last row/col length would result in a crash.
Fix: directly add last row/col len at tensor creation.

in update_from_top_indices() a number of top_indices smaller than a sparse block size would result in an empty mask and therefore crash the attention computation. This can happen for example on low % blending cases, because with low % the topk number of selected tokens can be smaller than a sparse block. E.g., with 0.0% 1 token is selected as a minimum.
Fix: have the mask have at least one block.

**What this PR does / why we need it**:

This PR fixes crashes that occur with FlashInfer sparse attention computation in edge cases where the sequence lengths are smaller than the sparse block mask, reason is explained above. This happens in blending when a low % of tokens are selected (or in the 0.0% case where a single token is selected by default).

**Special notes for your reviewers**:

Thank you for reviewing :)

**If applicable**:

- [ ] this PR contains user facing changes - docs added
- [ ] this PR contains unit tests

### CODE REVIEW

File: lmcache/v1/compute/attention/metadata.py [hunk 1]
```diff
@@ -71,7 +73,10 @@ def update_from_top_indices(self, top_indices: torch.Tensor):
         top_indices_slice = top_indices[
             self.sparse_blk_row_size - 1 :: self.sparse_blk_row_size
         ]
+        if len(top_indices_slice) == 0:
+            top_indices_slice = top_indices
```
[the remark this answers is not in the document]
Comment (rick-heig) [reply lmcache/v1/compute/attention/metadata.py]: top_indices is at least 1 element because because it comes from lmcache/v1/compute/blend/blender.py process_qkv() which samples topk_num top_indices from `top_indices = torch.topk(diff_k, k=topk_num).indices` which is guaranteed to be 1 by `topk_num = max(topk_num, 1)` and sequences of length 0 would not be blended as they would not get a cache hit, and even with a theoretical hit they would not pass the min blending tokens limit.

Anyways, it is reasonable to add a safety check here.

[the remark this answers is not in the document]
Comment (rick-heig) [reply lmcache/v1/compute/attention/metadata.py]: Updated the commit to include this.

**Comment by deng451e** [comment]
Default block size is 32. For shorter sequences, it’s probably more efficient to recompute fully instead of retrieving and updating the KV cache with blending.

**Comment by rick-heig** [comment]
> Default block size is 32. For shorter sequences, it’s probably more efficient to recompute fully instead of retrieving and updating the KV cache with blending.

@deng451e  I totally agree with you on that, but still, if for some reason a sequence shorter than 32 occurs, the whole system crashes, which is not great, especially in production.

I also know that I am testing/using edge case scenarios e.g., 0% blending (KV concatenation) which still recomputes 1 token, and therefore results in a crash as described above. It would be better to have it fixed anyways.

Edit: e.g., 0% blending on a 2000 token request is 1 token recompute, so a full recompute would not be only a sequence shorter than 32.

**Review by sammshen — APPROVED** [review]
LGTM

File: lmcache/v1/compute/attention/metadata.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (deng451e) [reply lmcache/v1/compute/attention/metadata.py]: Make sense.

**Review by deng451e — APPROVED** [review]

**Comment by rick-heig** [comment]
I had to update the PR as there was still an unhandled case that would lead to unaligned block masks and therefore a crash.

Below is the change I made to the PR:

```patch
--- a/lmcache/v1/compute/attention/metadata.py
+++ b/lmcache/v1/compute/attention/metadata.py
@@ -80,6 +80,11 @@ class LMCFlashInferSparseMetadata(LMCAttnMetadata):
             top_indices_slice = top_indices[-1:]
         top_indices_slice //= self.sparse_blk_col_size
         top_indices_slice = top_indices_slice.clamp_min(1) # ensure at least 1 block
+        if len(top_indices_slice) < len(block_row_sizes):
+            top_indices_slice = torch.cat([
+                top_indices_slice,
+                top_indices_slice[-1:].repeat(len(block_row_sizes) - len(top_indices_slice))
+            ])
         mask = cols < top_indices_slice.unsqueeze(1)
         block_mask_map[mask] = 1
         self.wrapper.plan
```

The PR code explicitly created a new block for the remainder so instead of having:

`block_row_sizes = [32,42]` it becomes `block_row_sizes = [32, 32, 10]` (because the current (old) code just adds the remainder value to last size with ` block_row_sizes[-1] += top_k_num % self.sparse_blk_row_size`
So now `len(block_row_sizes) = 3` but slicing still produces `len(top_indices_slice) = 2` which created a mismatch between `row_blocks = 3` and `slice_len = 2`. This leads to a crash in regular cases (sequences not smaller than block size).

The current (old) code had:
```
row_blocks = floor(top_k / row_size)
slice_len  = floor(top_k / row_size)
```

The new code has:
```
row_blocks = ceil(top_k / row_size)
slice_len  = floor(top_k / row_size)
```

Because slice_len still depends on

```python
        # NOTE(Jiayi): select every `sparse_blk_row_size`-th index from top_indices
        # to approximate the attention mask at block level.
        top_indices_slice = top_indices[
            self.sparse_blk_row_size - 1 :: self.sparse_blk_row_size
        ]
```

Which creates the case where `slice_len < row_blocks`, because the last element of the last partial block isn't covered by the "select every `sparse_blk_row_size`-th index from top_indices"

Therefore I added code to pad `top_indices_slice` with the last index, if necessary.

```patch
--- a/lmcache/v1/compute/attention/metadata.py
+++ b/lmcache/v1/compute/attention/metadata.py
@@ -80,6 +80,11 @@ class LMCFlashInferSparseMetadata(LMCAttnMetadata):
             top_indices_slice = top_indices[-1:]
         top_indices_slice //= self.sparse_blk_col_size
         top_indices_slice = top_indices_slice.clamp_min(1) # ensure at least 1 block
+        if len(top_indices_slice) < len(block_row_sizes):
+            top_indices_slice = torch.cat([
+                top_indices_slice,
+                top_indices_slice[-1:].repeat(len(block_row_sizes) - len(top_indices_slice))
+            ])
         mask = cols < top_indices_slice.unsqueeze(1)
         block_mask_map[mask] = 1
         self.wrapper.plan
```

Now all cases are covered.
Sorry for not testing this enough before.
I have now run it with multiple blending %, different sequence lengths, through many queries to make sure it works as expected.

Best,
Rick

File: lmcache/v1/compute/attention/metadata.py [hunk 2]
```diff
@@ -71,7 +73,15 @@ def update_from_top_indices(self, top_indices: torch.Tensor):
...
             self.sparse_blk_row_size - 1 :: self.sparse_blk_row_size
         ]
+        if len(top_indices_slice) == 0 and len(top_indices) > 0:
+            top_indices_slice = top_indices[-1:]
         top_indices_slice //= self.sparse_blk_col_size
+        top_indices_slice = top_indices_slice.clamp_min(1) # ensure at least 1 block
+        if len(top_indices_slice) < len(block_row_sizes):
+            top_indices_slice = torch.cat([
+                top_indices_slice,
+                top_indices_slice[-1:].repeat(len(block_row_sizes) - len(top_indices_slice))
```
Comment (sammshen) [inline lmcache/v1/compute/attention/metadata.py:83]: dumb question, what's the difference between `[-1]` and `[-1:]`

File: lmcache/v1/compute/attention/metadata.py [hunk 3]
```diff
@@ -71,7 +73,15 @@ def update_from_top_indices(self, top_indices: torch.Tensor):
         top_indices_slice = top_indices[
             self.sparse_blk_row_size - 1 :: self.sparse_blk_row_size
         ]
+        if len(top_indices_slice) == 0 and len(top_indices) > 0:
+            top_indices_slice = top_indices[-1:]
         top_indices_slice //= self.sparse_blk_col_size
+        top_indices_slice = top_indices_slice.clamp_min(1) # ensure at least 1 block
+        if len(top_indices_slice) < len(block_row_sizes):
```
Comment (sammshen) [inline lmcache/v1/compute/attention/metadata.py:80]: can you add some comments here

**Review by sammshen — CHANGES_REQUESTED** [review]
thanks for the follow up! could you detail the workloads / requests that trigger the edge cases?
